greenlight/lib/api/prefs.py

```diff
@@ -2,109 +2,174 @@
+
+            retval = self.marionette.execute_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+              let prefBranch = Services.prefs.QueryInterface(Ci.nsIPrefBranch);
 
-            ret = self.marionette.execute_script("""
-              let pref = arguments[0];
+              let pref_name = arguments[0];
               let value = arguments[1];
-              let prefBranch = Cc["@mozilla.org/preferences-service;1"]
-                               .getService(Ci.nsIPrefBranch);
-              let type = prefBranch.getPrefType(pref);
+
+              let type = prefBranch.getPrefType(pref_name);
+
+              // If the pref does not exist yet, get the type from the value
+              if (type == prefBranch.PREF_INVALID) {
+                switch (typeof value) {
+                  case "boolean":
+                    type = prefBranch.PREF_BOOL;
+                    break;
+                  case "number":
+                    type = prefBranch.PREF_INT;
+                    break;
+                  case "string":
+                    type = prefBranch.PREF_STRING;
+                    break;
+                  default:
+                    type = prefBranch.PREF_INVALID;
+                }
+              }
 
               switch (type) {
                 case prefBranch.PREF_BOOL:
-                  prefBranch.setBoolPref(pref, value);
+                  prefBranch.setBoolPref(pref_name, value);
                   break;
                 case prefBranch.PREF_STRING:
-                  prefBranch.setCharPref(pref, value);
+                  prefBranch.setCharPref(pref_name, value);
                   break;
                 case prefBranch.PREF_INT:
-                  prefBranch.setIntPref(pref, value);
+                  prefBranch.setIntPref(pref_name, value);
                   break;
                 default:
                   return false;
               }
 
               return true;
-            """, script_args=[pref, value])
+            """, script_args=[pref_name, value])
 
-        assert ret
+        assert retval
 
-    def restore_pref(self, pref):
-        """
-        Restore a previously set pref back to its original value.
+    def restore_all_prefs(self):
+        """Restore all previously set preferences to their original value."""
+        while len(self.archive):
+            self.restore_pref(self.archive.keys()[0])
+
+    def restore_pref(self, pref_name):
+        """Restore a previously set preference to its original value.
 
-        :param pref: The preference to restore. It must be a preference that
-                     was previously set using `set_pref`.
+        If the preference is not a default preference and has been newly
+        created, it will be removed.
+
+        :param name: The name of the preference to restore.
         """
-        return self.set_pref(pref, self.archive[pref])
+        assert pref_name
```

Comment:
Asserting arguments aren't falsey strikes me as a slightly odd pattern. I don't expect we'll ever find "", 0, [], {}, None, etc. in our pref archive. It's not really hurting anything, but it might be more straightforward to start this function with "if pref_name not in self.archive: raise ...".